src/idrs.jl

```diff
@@ -0,0 +1,183 @@
+        http://ta.twi.tudelft.nl/nw/users/gijzen/idrs.m
+    [4] IDR(s)' webpage http://ta.twi.tudelft.nl/nw/users/gijzen/IDR.html
+"""
+idrs(A, b; s = 8, tol=sqrt(eps(typeof(real(b[1])))), maxiter = length(b)^2) =
+    idrs_core!(zerox(A,b), linsys_op, (A,), b, s, tol, maxiter)
+
+idrs!(x, A, b; s = 8, tol=sqrt(eps(typeof(real(b[1])))), maxiter=length(x)^2) =
+    idrs_core!(x, linsys_op, (A,), b, s, tol, maxiter)
+
+function idrs_core!{T}(X, op, args, C::T, s, tol, maxiter)
+    R = C - op(X, args...)
+    normR = vecnorm(R)
+    res = [normR]
+	iter = 0                 
+
+    if normR <= tol           # Initial guess is a good enough solution
+        return X, ConvergenceHistory(0<= res[end] < tol, tol, length(res), res)
+    end
+    
+    Z = zero(C)
+
+    P = T[rand!(copy(C)) for k in 1:s]
+    U = T[copy(Z) for k in 1:s]
+    G = T[copy(Z) for k in 1:s]
+    Q = copy(Z)
+    V = copy(Z)
+    
+    M = eye(eltype(C),s,s)
+    f = zeros(eltype(C),s)
+    c = zeros(eltype(C),s)
+
+    om = 1.
+    iter = 0
+    while normR > tol && iter < maxiter
+        for i in 1:s,
+            f[i] = vecdot(P[i], R)
+        end
+        for k in 1:s 
+
+            # Solve small system and make v orthogonal to P
+
+            c = \(LowerTriangular(M[k:s,k:s]),f[k:s])
+            copy!(V, G[k])
+            scale!(c[1], V)
+        
+            copy!(Q, U[k])
+            scale!(c[1], Q)
+            for i = k+1:s
+                axpy!(c[i-k+1], G[i], V)
+                axpy!(c[i-k+1], U[i], Q)
+            end
+
+            # Compute new U[:,k] and G[:,k], G[:,k] is in space G_j
+
+            #V = R - V
+            scale!(-1., V)
+            axpy!(1., R, V)
+            
+            copy!(U[k], Q)
+            axpy!(om, V, U[k])
+            G[k] = op(U[k], args...) 
+
+            # Bi-orthogonalise the new basis vectors
+
+            for i in 1:k-1
+                alpha = vecdot(P[i],G[k])/M[i,i]
+                axpy!(-alpha, G[i], G[k])
+                axpy!(-alpha, U[i], U[k])
+            end
+
+            # New column of M = P'*G  (first k-1 entries are zero)
+
+            for i in k:s
+                M[i,k] = vecdot(P[i],G[k])       
+            end
+
+            #  Make r orthogonal to q_i, i = 1..k 
+
+            beta = f[k]/M[k,k]
+            axpy!(-beta, G[k], R)
+            axpy!(beta, U[k], X)
+        
+            normR = vecnorm(R)
+            res = [res; normR]
+            iter += 1
+            if normR < tol || iter > maxiter
+                return X, ConvergenceHistory(normR < tol, tol, length(res), res)
+            end 
+            if k < s 
+                f[k+1:s] = f[k+1:s] - beta*M[k+1:s,k]
+            end
+        
+        end
+
+        # Now we have sufficient vectors in G_j to compute residual in G_j+1
+        # Note: r is already perpendicular to P so v = r
+
+        copy!(V, R)
+        Q = op(V, args...)
+        om = omega(Q, R, 0.7)

```

Comment:
Maybe it doesn't really matter, but I think you can get the exact value for free because LLVM will compute the value as a constant
```
julia> f() = sqrt(2)/2;

julia> @code_llvm f()

; Function Attrs: sspreq
define double @julia_f_21376() #0 {
top:
  ret double 0x3FE6A09E667F3BCD
}
```